synapse/federation/sender/__init__.py

```diff
@@ -308,43 +331,68 @@ async def handle_event(event: EventBase) -> None:
                         destinations.discard(send_on_behalf_of)
 
-                    logger.debug("Sending %s to %r", event, destinations)
-
                     if destinations:
-                        await self._send_pdu(event, destinations)
-
                         now = self.clock.time_msec()
                         ts = await self.store.get_received_ts(event.event_id)
 
                         synapse.metrics.event_processing_lag_by_event.labels(
                             "federation_sender"
                         ).observe((now - ts) / 1000)
 
-                async def handle_room_events(events: Iterable[EventBase]) -> None:
-                    with Measure(self.clock, "handle_room_events"):
-                        for event in events:
-                            await handle_event(event)
-
-                events_by_room = {}  # type: Dict[str, List[EventBase]]
+                        return destinations
+                    return ()
+
+                async def get_federatable_events_and_destinations(
+                    events: Iterable[EventBase],
+                ) -> List[Tuple[EventBase, Collection[str]]]:
+                    with Measure(self.clock, "get_destinations_for_events"):
+                        # Fetch federation destinations per event,
+                        # skip if get_destinations_for_event returns an empty collection,
+                        # return list of event->destinations pairs.
+                        return [
+                            (event, dests)
+                            for (event, dests) in [
+                                (event, await get_destinations_for_event(event))
+                                for event in events
+                            ]
+                            if dests
+                        ]
+
+                events_by_room: Dict[str, List[EventBase]] = {}
                 for event in events:
                     events_by_room.setdefault(event.room_id, []).append(event)
 
-                await make_deferred_yieldable(
+                # concurrently process room events, to allow parallelization through db queries
+                nested_events_and_dests: List[
+                    List[Tuple[EventBase, Collection[str]]]
+                ] = await make_deferred_yieldable(
                     defer.gatherResults(
                         [
-                            run_in_background(handle_room_events, evs)
+                            run_in_background(
+                                get_federatable_events_and_destinations, evs
+                            )
                             for evs in events_by_room.values()
                         ],
                         consumeErrors=True,
                     )
                 )
 
+                # flatten list
+                events_and_dests = list(
+                    itertools.chain.from_iterable(nested_events_and_dests)
+                )
```

Comment:
I think it would be worth moving all this into a function with input `events` and output `events_and_dests`, so as to limit the scope of the intermediate variables `events_by_room` and `nested_events_and_dests`.